storage/src/adb/any.rs

```diff
@@ -78,18 +75,65 @@ pub struct Any<E: RStorage + Clock + Metrics, K: Array, V: Array, H: CHasher> {
     /// are over keys that have been updated by some operation at or after this point).
     inactivity_floor_loc: u64,
 
-    /// A map from each key to the location in the log containing its most recent update. Only
-    /// contains the keys that currently have a value (that is, deleted keys are not in the map).
-    snapshot: Index<EightCap, u64>,
+    /// A snapshot of all currently active operations in the form of a map from each key to the
+    /// location in the log containing its most recent update. Only contains the keys that currently
+    /// have a value (that is, deleted keys are not in the map).
+    snapshot: Index<T, u64>,
 
     /// The number of operations that are pending commit.
     uncommitted_ops: u64,
 }
 
-impl<E: RStorage + Clock + Metrics, K: Array, V: Array, H: CHasher> Any<E, K, V, H> {
-    /// Return an MMR initialized from `cfg`. Any uncommitted operations in the log will be
+/// The result of a database `update` operation.
+pub enum UpdateResult {
+    NoOp,              // tried to set a key to its current value
+    Inserted(u64),     // key was not previously in the snapshot & its new loc is the wrapped value.
+    Updated(u64, u64), // key was previously in the snapshot & its (old, new) loc pair is wrapped.
+}
+
+impl<E: RStorage + Clock + Metrics, K: Array, V: Array, H: CHasher, T: Translator>
+    Any<E, K, V, H, T>
+{
+    /// Returns any `Any` adb initialized from `cfg`. Any uncommitted log operations will be
     /// discarded and the state of the db will be as of the last committed operation.
-    pub async fn init(context: E, hasher: &mut H, cfg: Config) -> Result<Self, Error> {
+    pub async fn init(
+        context: E,
+        hasher: &mut H,
+        cfg: Config,
+        translator: T,
+    ) -> Result<Self, Error> {
+        let mut snapshot: Index<T, u64> = Index::init(context.with_label("snapshot"), translator);
+        let (mmr, log) = Self::init_mmr_and_log(context, hasher, cfg).await?;
+
+        let start_leaf_num = leaf_pos_to_num(mmr.pruned_to_pos()).unwrap();
+        let inactivity_floor_loc = Self::build_snapshot_from_log::<32 /* value is unused*/>(
```

Comment:
If not used, I recommend a const or 0.

Reply:
I tried to find an elegant way to remove this and couldn't 😢

Reply:
done